Made mmap-related code compatible with Cygwin

mvanleeu opened the pull request:
The whole logic of first blocking the memory + guard page, only to overwrite it afterwards in a fixed location does not work in cygwin (startp_ gets a value of 0xffffffffffff). So created a Cygwin path and simplified the mmap to a maximum.
Didn't bother to #ifdef all references to guardp_ though, which may have been cleaner.

Testing has been limited to the benchmark files, fitting in RAM.

mvanleeu committed:
Made mmap-related code compatible with Cygwin

The whole logic of first blocking the memory + guard page, only to overwrite it then in the fixed location does not work in cygwin (startp_ gets a value of 0xffffffffffff). So created a Cygwin path and simplified the mmap to a maximum.
Didn't bother to #ifdef all references to guardp_ though, which may have been cleaner.

Testing has been limited to the benchmark files, fitting in RAM.

mvanleeu committed:
Indicated cygwin compatibility

dw commented:
I haven't touched this code in a long time, but IIRC it is always necessary to be able to run 15 bytes over the end of the buffer, to allow pcmpistri to begin on the last byte (which might be necessary depending on the length of the file, and/or the prior data bytes appearing earlier in the buffer)

Won't merge this one without a bit more clarity, it looks like it might be breaking that requirement

mvanleeu committed:
test files for mmap

Verify if the use of SIMD instructions do not cause to read after the end of the mmap buffer.
Test files = length-{_n_}.csv with 'n' == 0 to 15 and representing the result of  file size(length-_n_.csv) modulo 16, as the SIMD instruction tested processes 16 bytes at a time.
Separate unittest file as the existing one somehow crashed.

mvanleeu committed:
Add files via upload

mvanleeu committed:
Merge branch 'dw:master' into master

mvanleeu commented:
Indeed, the description of csvmonkey mentions the approach of parsing 16 bytes at a time via SIMD. 
To see if I was just _lucky_ running the test in an ideal scenario, I have _doctored_ some small test data file that have slightly different lengths (i.e. "length of file" modulo 16 == "length-{_rest_value_}.csv", with the last field of the file filled with the character 'x'  equal to '_rest_value_ + 1' . I created test case to read and parse the last field of the file and verify if the number of 'x' characters matches.
Somehow, none of the length made the test program crashed or returned wrong values, which I would have expected if the mmap would cause the SMID instruction to process data that was after the end of the mmap.
Please review if this test case is actually testing the right thing.

mvanleeu commented:
```
$ python3 tests/csvmonkey_test_mmap.py
.
----------------------------------------------------------------------
Ran 1 test in 0.004s

OK
```

mvanleeu committed:
Fixes access violation under Windows 64 bits.
